diff -Naurd libapache-mod-perl-1.27.orig/apaci/mod_perl.config.sh libapache-mod-perl-1.27/apaci/mod_perl.config.sh
--- libapache-mod-perl-1.27.orig/apaci/mod_perl.config.sh	2002-05-23 05:20:54.000000000 +0200
+++ libapache-mod-perl-1.27/apaci/mod_perl.config.sh	2003-08-01 08:50:35.000000000 +0200
@@ -155,7 +155,7 @@
 
 print $ldopts;
 EOT
-perl_libs="`$perl_interp $tmpfile2 $build_type`"
+perl_libs="`$perl_interp $tmpfile2 $build_type $perl_libperl`"
 if test $build_type = OBJ
 then
 	case "$os_version" in
diff -Naurd libapache-mod-perl-1.27.orig/Makefile.PL libapache-mod-perl-1.27/Makefile.PL
--- libapache-mod-perl-1.27.orig/Makefile.PL	2002-06-02 04:13:25.000000000 +0200
+++ libapache-mod-perl-1.27/Makefile.PL	2003-08-01 08:54:38.000000000 +0200
@@ -530,6 +530,9 @@
 	    if (-e $lib) {
 		$LIBPERL = "-lperld";
 		$libperl = " -- $LIBPERL";
+	    } elsif (-e '/usr/lib/libdebugperl.a') { # Debian perl-debug
+	    	$LIBPERL = "-ldebugperl";
+		$libperl = " -- $LIBPERL";
 	    }
 	    $PERL_EXTRA_CFLAGS .= " -g";
 	    $PERL_TRACE=1;
